strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/ArtifactManagementService.java

```diff
@@ -141,12 +151,16 @@ public void releaseLock(URI pathUri)
 
     public void acquireLock(URI pathUri)
     {
-        Lock lock = Optional.ofNullable(pathUriMap.putIfAbsent(pathUri, lock = new ReentrantLock())).orElse(lock);
+        Lock lock = Optional.ofNullable(pathUriMap.putIfAbsent(pathUri,
+                                                               lock = hazelcastInstance.getLock(pathUri.toString())))
+                            .orElse(lock);
         lock.lock();
 
         while (!pathUriMap.containsKey(pathUri))
```

Comment:
@sbespalov : I know I raised this in the chat, but I think it would be great, if you could explain the purpose of this `while` loop here, as I'm not sure I'm getting it.

Reply:
this loop used to here to avoid "race donditions". For example two threads A and B can get same lock object, and then A accure lock first and remove this lock object from Map with `ArtifactManagementService.releaseLock`, so thread B will lock it after but other threads will not see this because there will be no this lock object in Map any more. So with this loop we guarantie that if thread accured lock then this lock object contains in Map and other threads can see this.